envoy/ssl/handshaker.h

```diff
@@ -156,5 +165,104 @@ class HandshakerFactory : public Config::TypedFactory {
   virtual SslCtxCb sslctxCb(HandshakerFactoryContext& handshaker_factory_context) const PURE;
 };
 
+struct SelectionResult {
+  enum class SelectionStatus {
+    // Continue the TLS handshake.
+    Continue,
+    // Block the TLS handshake.
+    Stop,
+    // Abort the TLS handshake.
+    AbortHandshake,
+  };
+  // If the value is Pending, the selection is asynchronous.
+  SelectionStatus status;
+  // It will be nullptr when status is Stop or AbortHandshake.
+  const Ssl::TlsContext* selected_ctx;
+  // Enable OCSP stapling response when it is true.
+  bool staple;
+};
+
+/**
+ * Used to return the result from an synchronous/asynchronous cert selection.
+ */
+class CertSelectionCallback {
+public:
+  virtual ~CertSelectionCallback() = default;
+
+  virtual Event::Dispatcher& dispatcher() PURE;
+
+  /**
+   * Called when the asynchronous cert selection completes.
+   * @param selected_ctx selected Ssl::TlsContext, it's empty when selection failed.
+   * @param staple true when need to set OCSP response.
+   */
+  virtual void onCertSelectionResult(OptRef<const Ssl::TlsContext> selected_ctx, bool staple) PURE;
+};
+
+using CertSelectionCallbackPtr = std::unique_ptr<CertSelectionCallback>;
+
+class TlsCertificateSelector {
+public:
+  virtual ~TlsCertificateSelector() = default;
+
+  /**
+   * select TLS context based on the client hello.
+   */
+  virtual SelectionResult selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello,
+                                           CertSelectionCallbackPtr cb) PURE;
+};
+
+using TlsCertificateSelectorPtr = std::unique_ptr<TlsCertificateSelector>;
+
+class ContextSelectionCallback {
+public:
+  virtual ~ContextSelectionCallback() = default;
+
+  /**
+   * @return reference to the initialized Tls Contexts.
+   */
+  virtual const std::vector<TlsContext>& getTlsContexts() const PURE;
+};
+
+using ContextSelectionCallbackWeakPtr = std::weak_ptr<ContextSelectionCallback>;
+
+using TlsCertificateSelectorFactoryCb =
+    std::function<TlsCertificateSelectorPtr(ContextSelectionCallbackWeakPtr)>;
+
+class TlsCertificateSelectorFactoryContext {
+public:
+  virtual ~TlsCertificateSelectorFactoryContext() = default;
+
+  /**
+   * Returns the singleton manager.
+   */
+  virtual Singleton::Manager& singletonManager() PURE;
+
+  /**
+   * @return reference to the server options
+   */
+  virtual const Server::Options& options() const PURE;
+
+  /**
+   * @return reference to the Api object
+   */
+  virtual Api::Api& api() PURE;
+};
+
+class TlsCertificateSelectorFactory : public Config::TypedFactory {
+public:
+  /**
+   * @returns a callback to create a TlsCertificateSelector. Accepts the |config| and
+   * |validation_visitor| for early validation. This virtual base doesn't
+   * perform MessageUtil::downcastAndValidate, but an implementation should.
+   */
+  virtual TlsCertificateSelectorFactoryCb createTlsCertificateSelectorCb(
+      const ProtobufWkt::Any& config,
```

Comment:
Can this be a `Protobuf::Message&` instead? Similar to other factories, such as HandshakerFactory. See related code in `context_config_impl.cc` near line 242